background-processor/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+				if bytes == expected_bytes {
+					done_persisting = true;
+				}
+			}
+		}
+
+		// Check that the initial channel manager data is persisted as expected.
+		let filepath_0 = get_full_filepath("test_background_processor_persister_0".to_string(), "manager".to_string());
+		let filepath_1 = get_full_filepath("test_background_processor_persister_1".to_string(), "manager".to_string());
+		loop {
+			check_persisted_data!(nodes[0].node, filepath_0.clone());
+			if done_persisting {
+				// Check that eventually BackgroundProcessor resets the condvar when everything's done persisting.
+				loop {
+					if !nodes[0].node.get_persistence_condvar_value() { break }
+				}
+				break
+			}
+		}
+		done_persisting = false;
+		loop {
+			check_persisted_data!(nodes[1].node, filepath_1.clone());
+			if done_persisting {
+				loop {
+					if !nodes[1].node.get_persistence_condvar_value() { break }
+				}
+				break
+			}
+		}
+
+		// Force-close the channel.
+		nodes[0].node.force_close_channel(&OutPoint { txid: tx.txid(), index: 0 }.to_channel_id());
+
+		// Check that the force-close updates are persisted.
+		done_persisting = false;
+		loop {
+			check_persisted_data!(nodes[0].node, filepath_0.clone());
+			if done_persisting {
+				loop {
+					if !nodes[0].node.get_persistence_condvar_value() { break }
+				}
+				break
+			}
+		}
+	}
+
+	#[test]
+	fn test_chan_freshness_called() {
+		// Test that ChannelManager's `timer_chan_freshness_every_min` is called every
+		// `CHAN_FRESHNESS_TIMER`.
+		let nodes = create_nodes(1, "test_chan_freshness_called".to_string());
+		let data_dir = nodes[0].persister.get_data_dir();
+		let callback = move |node| FilesystemPersister::persist_manager(data_dir.clone(), node);
+		let processor = BackgroundProcessor::start(callback, nodes[0].node.clone(), nodes[0].logger.clone());
+		loop {
+			let log_entries = nodes[0].logger.lines.lock().unwrap();
+			let desired_log = "Calling manager's timer_chan_freshness_every_min".to_string();
+			if log_entries.get(&("background_processor".to_string(), desired_log)).is_some() {
+				break
+			}
+		}
+		processor.stop();
+		processor.thread_handle.join().unwrap();
+	}
+
+	#[test]
+	fn test_persist_error() {
+		// Test that if we encounter an error during manager persistence, the thread panics.
+		fn persist_manager<ChanSigner, M, T, K, F, L>(_data: Arc<ChannelManager<ChanSigner, Arc<M>, Arc<T>, Arc<K>, Arc<F>, Arc<L>>>) -> Result<(), std::io::Error>
+		where ChanSigner: 'static + ChannelKeys + Writeable,
+		      M: 'static + chain::Watch<Keys=ChanSigner>,
+		      T: 'static + BroadcasterInterface,
+		      K: 'static + KeysInterface<ChanKeySigner=ChanSigner>,
+		      F: 'static + FeeEstimator,
+		      L: 'static + Logger,
+		{
+			Err(std::io::Error::new(std::io::ErrorKind::Other, "test"))
+		}
+
+		let nodes = create_nodes(2, "test_persist_error".to_string());
+		let bg_processor = BackgroundProcessor::start(persist_manager, nodes[0].node.clone(), nodes[0].logger.clone());
```

Comment:
It is somewhat awkward for the tests to print that a panic occurred and it appear on the console despite the test passing - is there some way to capture the error printing in the background thread?